SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/Components/SpatialPingComponent.h"
+
+#include "GameFramework/PlayerController.h"
+#include "Net/UnrealNetwork.h"
+#include "TimerManager.h"
+#include "Engine/World.h"
+
+USpatialPingComponent::USpatialPingComponent(const FObjectInitializer& ObjectInitializer) : Super(ObjectInitializer)
+{
+	PrimaryComponentTick.bCanEverTick = false;
+	PrimaryComponentTick.bStartWithTickEnabled = false;
+
+	bReplicates = true;
+
+	bStartWithPingEnabled = true;
+}
+
+void USpatialPingComponent::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(USpatialPingComponent, ReplicatedPingID);
+}
+
+void USpatialPingComponent::BeginPlay()
+{
+	Super::BeginPlay();
+	//Attempt to cast the owner of this component to the PlayerController class
+	//Component will do nothing if cast fails
+	OwningController = Cast<APlayerController>(GetOwner());
+	if (bStartWithPingEnabled)
+		SetPingEnabled(true);
+}
+
+void USpatialPingComponent::EndPlay(const EEndPlayReason::Type EndPlayReason)
+{
+	Super::EndPlay(EndPlayReason);
+	//Clear up the timer before the component is disabled or destroyed
+	SetPingEnabled(false);
+}
+
+bool USpatialPingComponent::GetIsPingEnabled() const
+{
+	return bIsPingEnabled;
+}
+
+void USpatialPingComponent::SetPingEnabled(bool bSetEnabled)
+{
+	if (OwningController) {
+		//Only execute on owning local client
+		if (OwningController->GetLocalRole() == ENetRole::ROLE_AutonomousProxy)
+		{
+			if (bSetEnabled && !bIsPingEnabled) {
+				EnablePing();
+			}
+			else if (!bSetEnabled && bIsPingEnabled) {
+				DisablePing();
+			}
+		}
+	}
+}
+
+float USpatialPingComponent::GetPing() const
+{
+	return RTPing;
+}
+
+void USpatialPingComponent::EnablePing()
+{
+	UWorld* World = GetWorld();
+	if (World)
+	{
+		//Set looping timer to 'tick' this component and send ping RPC with configurable frequency
+		World->GetTimerManager().SetTimer(PingTimerHandle, this, &USpatialPingComponent::TickPingComponent, PingFrequency, true);
+		bIsPingEnabled = true;
+	}
+}
+
+void USpatialPingComponent::DisablePing()
+{
+	UWorld* World = GetWorld();
+	if (World)
+	{
+		//Clear the timer
+		World->GetTimerManager().ClearTimer(PingTimerHandle);
+		bIsPingEnabled = false;
+		RTPing = 0.f;
+	}
+}
+
+void USpatialPingComponent::TickPingComponent()
+{
+	SendNewPing();
+	//Pass latest measured ping to owning controller to be processed by PlayerState
+	OwningController->UpdatePing(RTPing);
+}
+
+void USpatialPingComponent::SendNewPing()
+{
+	//Send a new ping using the current local time since start as both ID and timestamp
+	SendServerWorkerPingID(GetWorld()->GetRealTimeSeconds());
```

Comment:
Why does this need a time? Surely you need to be able to match a ping to a pong, and the times for both should be stored locally.

Reply:
Also why is it called ID?

Reply:
Since the timestamp was a 32 bit float, the same as the IDs, I came to the conclusion that just sending the timestamp would be a lighter implementation without any need for finding a matching ping on receiving a pong. If the size of the timestamp were larger then yes, I'd opt to store them locally and match the pairs.

Am I overlooking something?

Reply:
It's harder to work out what is being done if there is an actual timestamp but I guess that could be commented.
The bigger issue is what accuracy do you need on this. I estimate that you lose millisecond accuracy on that timestamp after a few hours.

Reply:
Yeah you're correct on the accuracy point. I'll cache the times locally in higher resolution then.